internal/pkg/client/cache/dir.go

```diff
@@ -22,69 +22,251 @@ const (
+	os.Setenv(DirEnv, baseDir)
+	newCache.PreviousDirEnv = savedDirEnv
+
+	newCache.BaseDir = baseDir
+	isDefaultCache, err := isDefaultBasedir(baseDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if this is the default cache: %s", err)
+	}
+	newCache.Root = rootDir
+	newCache.State = StateInitialized
+	newCache.Default = isDefaultCache
+	newCache.Library, err = getLibraryCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Library cache: %s", err)
+	}
+	newCache.OciTemp, err = getOciTempCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the OCI temp cache")
+	}
+	newCache.OciBlob, err = getOciBlobCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the OCI blob cache")
+	}
+	newCache.Net, err = getNetCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Net cache")
+	}
+	newCache.Shub, err = getShubCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Shub cache")
+	}
 
-// Root is the root location where all of singularity caching happens. Library, Shub,
-// and oci image formats supported by containers/image repository will be cached inside
-//
-// Defaults to ${HOME}/.singularity/cache
-func Root() string {
-	updateCacheRoot()
+	// Sanity check to ensure that everything is fine before returning the
+	// handle. We do not know a way at the moment to reach the error case.
+	if !newCache.IsValid() {
+		return nil, fmt.Errorf("unable to correctly initialize new cache")
+	}
 
-	return root
+	return newCache, nil
 }
 
-// Clean : wipes all files in the cache directory, will return a error if one occurs
-func Clean() error {
-	sylog.Debugf("Removing: %v", Root())
+// Destroy a specific Singularity cache
+func (c *SingularityCache) Destroy() error {
+	sylog.Debugf("Removing: %v", c.Root)
+	if !c.IsValid() {
+		return fmt.Errorf("invalid cache")
+	}
 
-	if err := os.RemoveAll(Root()); err != nil {
-		return fmt.Errorf("unable to clean all cache: %s", err)
+	err := os.RemoveAll(c.Root)
+	if err != nil {
+		return fmt.Errorf("failed to delete the cache: %s", err)
 	}
 
 	return nil
 }
 
-func updateCacheRoot() {
+// IsValid checks whether a given Singularity cache is valid or not
+func (c *SingularityCache) IsValid() bool {
+	if c.State != StateInitialized {
+		return false
+	}
+
+	if c.BaseDir == "" {
+		return false
+	}
+
+	// Since Clean/Destroy delete everything in the cache directory,
+	// we make sure that when the user set the environment variable
+	// to specify where the cache should be, it cannot be in critical
+	// directory such as $HOME
+	usr, err := user.Current()
+	if err != nil {
+		return false
+	}
+
+	// The root cannot be empty of the Home directory. If the root was
+	// to be the home directory, destroying the cache would delete
+	// user's data
+	if c.Root == "" || c.Root == usr.HomeDir {
```

Comment:
this couples the notion of validity with the user of the cache.

Why would a cache be _invalid_ if it's in the user's home directory? My guess is that the problem is not the function, but the name of this function. `IsTransient`, maybe?

Reply:
I have to disagree with this since the current semantic of clean is to delete the root. If it is the user directory, we would delete their home (it happens to me, I misunderstood the semantic of `Clean()` at first). So we either change the semantic of `Clean()` or keep the proposed modification. I think simply switching to `IsTransient()` would be extremely dangerous.

Reply:
I am sorry but it does not make sense to me. Root is designed to be a way to protect the code to do bad thing and I read here that it is not a valuable feature:
- `Init` already refuses to create the cache is the home directory is used (L#141-142),
- it is possible, once the cache created to programmatically change the root (`myCache.Root = myHomeDir`), which is why a `IsValid()` function is so necessary: check that everything is fine before performing any action with the handle. While testing I had a few problems related to this and I do not feel confortable not including so a feature.

I am going to pull out my PR, I do not feel comfortable removing the protection that `Root` vs `Basedir` offers and do not see how I could make things better. Sorry.

As to not export `Root`, I agree with your point.